Refactor solana_swap_validator_hosts_v2 role for RBAC-enabled hosts

MexicanSparker opened the pull request:
# Refactor `solana_swap_validator_hosts_v2` role for RBAC-enabled hosts

## 📝 Summary

This PR refactors the `solana_swap_validator_hosts_v2` role to properly support RBAC-enabled validator hosts, addressing fundamental differences in user model, directory structure, file permissions, and SSH key management between legacy and RBAC environments. The role now correctly handles operator user accounts, selective privilege escalation, and proper SSH authentication for host-to-host communication.

## 🎯 Type of Change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [ ] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation update
- [ ] 🔧 Configuration change
- [ ] 🚀 Performance improvement
- [x] ♻️ Code refactoring (no functional changes)

## 🔍 Scope & Complexity

**Please keep PRs focused and small for faster reviews.**

- [x] This PR changes **fewer than 400 lines** of code (excluding generated files)
- [x] This PR addresses **only one logical change** or feature
- [ ] This PR can be **reviewed in under 30 minutes**
- [x] This PR does **not mix multiple types of changes** (e.g., refactoring + new features)

## 📋 Changes Made

Detailed list of changes:

- **Playbook user model**: Changed from `solana_user` to `operator_user` in `pb_hot_swap_validator_hosts_v2.yml` to match the pattern used in `pb_install_solana_cli_agave.yml`, ensuring operators run tasks as their own accounts
- **Directory paths**: Updated playbook pre_tasks to use `validator_key_store` (V2 path `/opt/validator/keys`) instead of `keys_dir` (legacy path `/home/sol/keys`)
- **Privilege escalation**: Added `become: true` to tasks that need to read the keys directory, following the pattern of `become: false` at play level with selective escalation
- **RPC URL handling**: Added `RPC_URL` environment variable to `ansible-control-localnet` container in docker-compose.yml to fix empty URL parameter errors
- **SSH key management**: 
  - Changed SSH key authorization from `sol` user to `ansible_user` (operator user) in both `prepare.yml` and `cleanup_ssh_keys.yml`
  - SSH keys are now generated in operator's home directory and authorized for operator accounts, not the `sol` service user
- **Tower file transfer**:
  - Fixed rsync command to use operator's SSH key with proper options to bypass host key verification
  - Added known_hosts cleanup before transfer
  - Added SSH connection test before rsync attempt
  - Improved error handling to stop playbook on failure
- **Validator command execution**: Added `become: true` and `become_user: "{{ solana_user }}"` to all `agave-validator` commands (wait-for-restart-window, set-identity) since these require access to ledger files that only the `sol` user can access

## 🧪 Testing

- [ ] Pre-commit hooks pass (`pre-commit run --all-files`)
- [x] Manual testing completed
- [x] Existing functionality verified unchanged
- [ ] Documentation updated (if applicable)

### Test Instructions

## 1. Deploy users
```bash
# Deploy users to host-bravo
ansible-playbook playbooks/pb_setup_users_validator.yml \
  -i solana_localnet.yml \
  -e "target_host=host-bravo" \
  -e "ansible_user=ubuntu" \
  -e "csv_file=iam_setup_dev.csv"

# Deploy users to host-charlie
ansible-playbook playbooks/pb_setup_users_validator.yml \
  -i solana_localnet.yml \
  -e "target_host=host-charlie" \
  -e "ansible_user=ubuntu" \
  -e "csv_file=iam_setup_dev.csv"
```

## 2. Setup user password for bob in host-charlie and host-delta
```bash
# Log in into host-bravo
ssh bob@host-bravo
sudo reset-my-password

# Log in into host-charlie
ssh bob@host-charlie
sudo reset-my-password
```

## 3. Install the primary validator using `demo2` keyset

### Install an Agave validator from prcompiled binaries on host-bravo as primary target
```bash
ansible-playbook playbooks/pb_setup_validator_agave.yml \
  -i solana_localnet.yml \
  --limit host-bravo \
  -e "target_host=host-bravo" \
  -e "ansible_user=bob" \
  -e "validator_name=demo2" \
  -e "validator_type=primary" \
  -e "agave_version=2.3.11" \
  -e "solana_cluster=localnet" \
  -e "build_from_source=false" \
  -K
```

## 4. Fund demo2 primary identity and add stake to demo2 vote account

To fund the primary identity for the `demo2` keyset and add 200k SOL stake to the validator's vote account execute these sequence of commands:

```bash
# Move into the demo2 keyset directory
cd /hayek-validator-kit/solana-localnet/validator-keys/demo2

# Set working cluster in solana config
solana config set --url http://localhost:8899

# Airdrop some SOL to the identity account
solana airdrop 40 primary-target-identity.json

# Create the validator vote account
solana create-vote-account vote-account.json primary-target-identity.json authorized-withdrawer.json

# Create a stake account with 200k SOL
solana create-stake-account stake-account.json 200000

# Delegate stake to the validator's vote account
solana delegate-stake stake-account.json vote-account.json --force
```

At this point wait some minutes to verify that demo2 validator `demo2n4nr6GHdsjF8pQco7BMp7ahy4mEcBg1Y6uaN5q` is voting (non-delinquent)

```bash
ansible-control:/hayek-validator-kit/ansible$ solana gossip -ul
IP Address      | Identity                                     | Gossip | TPU   | TPU-QUIC | RPC Address           | Version | Feature Set
----------------+----------------------------------------------+--------+-------+----------+-----------------------+---------+----------------
172.25.0.10     | 2mYS4raNeG8nymvbY55XvmVNcpkGiVz26WQKEJsNdm4r | 8001   | 8003  | 8009     | 172.25.0.10:8899      | 2.2.20  | 3073396398
172.25.0.11     | demoneTKvfN3Bx2jhZoAHhNbJAzt2rom61xyqMe5Fcw  | 8000   | 8003  | 8009     | 172.25.0.11:8899      | 2.2.20  | 3073396398
172.25.0.12     | demo2n4nr6GHdsjF8pQco7BMp7ahy4mEcBg1Y6uaN5q  | 8001   | 8003  | 8009     | none                  | 2.3.11  | 2142755730
Nodes: 3

ansible-control:/hayek-validator-kit/ansible$ solana validators -ul --keep-unstaked-delinquents
   Identity                                      Vote Account                            Commission  Last Vote        Root Slot     Skip Rate  Credits  Version            Active Stake
  demoneTKvfN3Bx2jhZoAHhNbJAzt2rom61xyqMe5Fcw   demo52s9s1foFXgnbVa8vYQM8GS9XRsJ3aMpus1rNnb   100%       4532 (  0)       4501 (  0)   0.00%      544   2.2.20    199999.997717120 SOL (14.27%)
  demo2n4nr6GHdsjF8pQco7BMp7ahy4mEcBg1Y6uaN5q   demo2uwCnuc8HvTTfHYd6Eq714d4XcqE1cTffRZcGru   100%       4532 (  0)       4501 (  0)    -         544   2.3.11    199999.997717120 SOL (14.27%)
  2mYS4raNeG8nymvbY55XvmVNcpkGiVz26WQKEJsNdm4r  EX3eoAibXd4nHXbXq1En8E6HXmf1hB2rVpdN4XeoRvnA    0%       4532 (  0)       4501 (  0)   0.00%      544   2.2.20   1001919.644800378 SOL (71.47%)

Average Stake-Weighted Skip Rate: 0.00%
Average Unweighted Skip Rate:     0.00%

Active Stake: 1401919.640234618 SOL

Stake By Version:
2.3.11  -    1 current validators (14.27%)
2.2.20  -    2 current validators (85.73%)
```

## 5. Install the hot-spare validator using `demo2` keyset

### Install an Agave validator from prcompiled binaries on host-charlie as hot-spare
```bash
ansible-playbook playbooks/pb_setup_validator_agave.yml \
  -i solana_localnet.yml \
  --limit host-charlie \
  -e "target_host=host-charlie" \
  -e "ansible_user=bob" \
  -e "validator_name=demo2" \
  -e "validator_type=hot-spare" \
  -e "agave_version=2.3.11" \
  -e "solana_cluster=localnet" \
  -e "build_from_source=false" \
  -K
```

### Verify that the hot-spare validator started up nicely and that it appears in gossip
```bash
ansible-control:/hayek-validator-kit/ansible$ solana gossip -ul
IP Address      | Identity                                     | Gossip | TPU   | TPU-QUIC | RPC Address           | Version | Feature Set
----------------+----------------------------------------------+--------+-------+----------+-----------------------+---------+----------------
172.25.0.10     | 2mYS4raNeG8nymvbY55XvmVNcpkGiVz26WQKEJsNdm4r | 8001   | 8003  | 8009     | 172.25.0.10:8899      | 2.2.20  | 3073396398
172.25.0.11     | demoneTKvfN3Bx2jhZoAHhNbJAzt2rom61xyqMe5Fcw  | 8000   | 8003  | 8009     | 172.25.0.11:8899      | 2.2.20  | 3073396398
172.25.0.12     | demo2n4nr6GHdsjF8pQco7BMp7ahy4mEcBg1Y6uaN5q  | 8001   | 8003  | 8009     | none                  | 2.3.11  | 2142755730
172.25.0.13     | 9Bz8siBR3vLwA3L6VPjBdJ5f5EobHG2krUGfizyv2yUp | 8001   | none  | none     | none                  | 2.3.11  | 2142755730
Nodes: 4
```

## 6. Run hot_swap_validator_hosts

### Swap validator hosts host-bravo -> host-charlie
```bash
ansible-playbook playbooks/pb_hot_swap_validator_hosts_v2.yml \
  -i solana_localnet.yml \
  -e "operator_user=bob" \
  -e "source_host=host-bravo" \
  -e "destination_host=host-charlie" \
  -K
```

Verify with gossip that the primary identity is now running at `172.25.0.13`
```bash
ansible-control:/hayek-validator-kit/ansible$ solana gossip -ul
IP Address      | Identity                                     | Gossip | TPU   | TPU-QUIC | RPC Address           | Version | Feature Set
----------------+----------------------------------------------+--------+-------+----------+-----------------------+---------+----------------
172.25.0.10     | Go8T8QcgQJePhasJih2nFQAoMPrcWYiyYnQR5qZbweKQ | 8001   | 8003  | 8009     | 172.25.0.10:8899      | 2.2.20  | 3073396398
172.25.0.11     | demoneTKvfN3Bx2jhZoAHhNbJAzt2rom61xyqMe5Fcw  | 8000   | 8003  | 8009     | 172.25.0.11:8899      | 2.2.20  | 3073396398
172.25.0.13     | demo2n4nr6GHdsjF8pQco7BMp7ahy4mEcBg1Y6uaN5q  | 8001   | 8003  | 8009     | none                  | 2.3.11  | 2142755730
172.25.0.12     | 65bpwuefNcvB9vbKC6zD3vr6cVKRRHg5hcuWrRQg71jP | 8001   | 8003  | 8009     | none                  | 2.3.11  | 2142755730
Nodes: 4
```

### Swap validator hosts host-charlie -> host-bravo (optional)

```bash
ansible-playbook playbooks/pb_hot_swap_validator_hosts_v2.yml \
  -i solana_localnet.yml \
  -e "operator_user=bob" \
  -e "source_host=host-charlie" \
  -e "destination_host=host-bravo" \
  -K
```

## 📚 Documentation

- [ ] Updated relevant documentation
- [ ] Added/updated comments for complex logic
- [ ] README updated (if applicable)
- [x] No documentation changes needed

**Note:** Documentation updates will be handled in a separate PR.

## 🔗 Related Issues

N/A

## 📝 Review Notes

Any specific areas you'd like reviewers to focus on or context they should know:

**Key Architectural Changes:**
1. **Selective Privilege Escalation**: The playbook follows a pattern of `become: false` at the play level, with `become: true` only on specific tasks that need it:
   - Tasks accessing `/opt/validator/keys` directory
   - Tasks executing `agave-validator` commands (require `sol` user for ledger access)

2. **SSH Key Flow**: 
   - Keys generated in operator's home directory on source host
   - Public key authorized for operator user on destination host
   - This allows operator-to-operator SSH, not operator-to-sol

3. **Tower File Transfer**: The rsync command uses the operator's SSH key with options to bypass host key verification, since operators may not have previously connected between these specific hosts.

---

## For Reviewers

**Estimated review time:** ⏱️ [45-60 minutes]

**Focus areas:**
- [x] Logic correctness
- [x] Security implications (especially privilege escalation and SSH key handling)
- [ ] Performance impact
- [ ] Documentation completeness

MexicanSparker commented on ansible/roles/solana_swap_validator_hosts_v2/tasks/prepare.yml:
```diff
@@ -203,7 +219,7 @@
+    PATH: "{{ system_solana_active_release_dir }}"
   when: inventory_hostname == source_host
   changed_when: false
   check_mode: false
```
Solved on commit https://github.com/team-supersafe/hayek-validator-kit/pull/165/commits/0c7dc6c18a5f52053906a40f0f92abf77c9f3070

eydelrivero commented:
@MexicanSparker The swap fails. I've tried 3 times...

```sh
ansible-playbook playbooks/pb_hot_swap_validator_hosts_v2.yml \
  -i solana_localnet.yml \
  -e "operator_user=bob" \
  -e "source_host=host-bravo" \
  -e "destination_host=host-charlie" \
  -K
```

<img width="1135" height="829" alt="image" src="https://github.com/user-attachments/assets/1bee01fe-2c58-4e98-8090-cbf872b193c5" />

MexicanSparker commented:
> @MexicanSparker Swap still not working
> 
> image
> ... and yet, the playbook ignores the failure and continues, leaving both host running the non primary identity
> 
> image

**Cause:** The error was a race condition where the destination validator (agave-validator) had not finished initializing its RPC interface before receiving the `set-identity` command.

**Solution:**
1.  **Robust Pre-check:** I implemented a strict health check which runs `solana catchup --our-localhost`. This guarantees the destination validator is fully synced and RPC-ready *before* any swap logic begins.
2.  **Atomic Failure:** I added `any_errors_fatal: true` to critical blocks. This ensures that if any check or task fails on any host, the entire operation stops immediately for all hosts, preventing inconsistent states.

eydelrivero commented:
@MexicanSparker still no dice

<img width="1251" height="609" alt="image" src="https://github.com/user-attachments/assets/e965292a-e232-4613-952a-3633015db110" />

MexicanSparker commented:
> Running the swap playbook with the source host running the hot-spare identity and the destination host already running the primary identity goes all the way through and even detects missed credits when in reality it didn't do anything. This should short-circuit and stop when it detects that necessary conditions for the swap are not met.

It turns out that I had a check for this purpose but somehow along the short lifespan of this role the conditions for the failure was modified I believe accidentally and this is on me. I have rectified the condition and the messaging of this fail task. Now it correctly prevents attempting to swap on hosts already in the desired state.
